File: codes/java/chap08/answer_A59.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 入力
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int Q = sc.nextInt();
		int[] queryType = new int[Q + 1];
		int[] pos = new int[Q + 1];
		int[] x = new int[Q + 1];
		int[] l = new int[Q + 1];
		int[] r = new int[Q + 1];
		for (int i = 1; i <= Q; i++) {
			queryType[i] = sc.nextInt();
			if (queryType[i] == 1) {
				pos[i] = sc.nextInt();
				x[i] = sc.nextInt();
			}
			if (queryType[i] == 2) {
				l[i] = sc.nextInt();
				r[i] = sc.nextInt();
			}
		}

		// クエリの処理
		SegmentTree Z = new SegmentTree(N);
		for (int i = 1; i <= Q; i++) {
			if (queryType[i] == 1) {
				Z.update(pos[i], x[i]);
			}
			if (queryType[i] == 2) {
				// 最初のセルに対応する半開区間は [1, size + 1)
				System.out.println(Z.query(l[i], r[i], 1, Z.size + 1, 1));
			}
		}
	}

	// セグメント木を実装したクラス SegmentTree
	static class SegmentTree {
		int size;
		int[] dat;
		// 要素 dat の初期化を行う（最初は全部ゼロ）
		SegmentTree(int n) {
			size = 1;
			while (size < n) {
				size *= 2;
			}
			dat = new int[size * 2]; // Java では new で初期化した配列の要素は 0 になることに注意
		}
		// クエリ 1 に対する処理
		void update(int pos, int x) {
			pos += size - 1;
			dat[pos] = x;
			while (pos >= 2) {
				pos /= 2;
				dat[pos] = dat[pos * 2] + dat[pos * 2 + 1]; // 8.8 節から変更した部分
			}
		}
		// クエリ 2 に対する処理
		// u は現在のセル番号、[a, b) はセルに対応する半開区間、[l, r) は求めたい半開区間
		int query(int l, int r, int a, int b, int u) {
			if (r <= a || b <= l) {
				return 0; // 8.8 節から変更した部分
			}
			if (l <= a && b <= r) {
				return dat[u];
			}
			int m = (a + b) / 2;
			int answerL = query(l, r, a, m, u * 2);
			int answerR = query(l, r, m, b, u * 2 + 1);
			return answerL + answerR; // 8.8 節から変更した部分
		}
	}
}

File: codes/python/chap08/answer_A59.py
# セグメント木（ここでは書籍とは異なり、pos が 0-indexed になるように実装しています）
class segtree:
	# 要素 dat の初期化を行う（最初は全部ゼロ）
	def __init__(self, n):
		self.size = 1
		while self.size < n:
			self.size *= 2
		self.dat = [ 0 ] * (self.size * 2)
	
	# クエリ 1 に対する処理
	def update(self, pos, x):
		pos += self.size # pos は 0-indexed なので、A[pos] のみに対応するセルの番号は pos + size
		self.dat[pos] = x
		while pos >= 2:
			pos //= 2
			self.dat[pos] = self.dat[pos * 2] + self.dat[pos * 2 + 1] # 8.8 節から変更した部分
	
	# クエリ 2 に対する処理
	# u は現在のセル番号、[a, b) はセルに対応する半開区間、[l, r) は求めたい半開区間
	def query(self, l, r, a, b, u):
		if r <= a or b <= l:
			return 0 # 8.8 節から変更した部分
		if l <= a and b <= r:
			return self.dat[u]
		m = (a + b) // 2
		answerl = self.query(l, r, a, m, u * 2)
		answerr = self.query(l, r, m, b, u * 2 + 1)
		return answerl + answerr # 8.8 節から変更した部分

# 入力
N, Q = map(int, input().split())
queries = [ list(map(int, input().split())) for i in range(Q) ]

# クエリの処理
Z = segtree(N)
for q in queries:
	tp, *cont = q
	if tp == 1:
		pos, x = cont
		Z.update(pos - 1, x) # pos は 1-indexed で入力されるので、update 関数の引数は pos - 1 にします
	if tp == 2:
		l, r = cont
		answer = Z.query(l - 1, r - 1, 0, Z.size, 1) # 0-indexed の実装では、最初のセルに対応する半開区間は [0, size) です
		print(answer)